examples/source_examples/accelo_api_connector_multithreading_enabled/connector.py

```diff
@@ -0,0 +1,828 @@
+
+        with ThreadPoolExecutor(max_workers=MAX_WORKERS) as executor:
+            page = 0
+            while True:
+                futures = [executor.submit(fetch_page, p) for p in range(page, page + MAX_WORKERS)]
+                results = [future.result() for future in futures]
+                all_entities = [entity for page_entities in results for entity in page_entities if page_entities]
+
+                for entity in all_entities:
+                    if process_record:
+                        entity = process_record(entity)
+                    if entity:
+                        yield op.upsert(entity_name, {field: entity.get(field) for field in fields})
+                        records_processed += 1
+                        if records_processed % 100 == 0:
+                            log.info(f"Processed {records_processed} records for {entity_name}")
+
+                if not any(results) or time.time() - start_time >= timeout:
+                    break
+                page += MAX_WORKERS
+
+        new_last_sync = datetime.fromtimestamp(max(max_date_value, int(time.time())), tz=timezone.utc).strftime(
+            '%Y-%m-%dT%H:%M:%SZ')
+        thread_local_state.state[last_sync_key] = new_last_sync
+        log.info(
+            f"Sync completed for {entity_name}. Entities received: {entities_received}, Records processed: {records_processed}. Time taken: {time.time() - start_time:.2f} seconds")
+
+    except Exception as e:
+        log.severe(f"Unhandled exception during sync of {entity_name}: {str(e)}")
+        log.severe(f"Exception details: {type(e).__name__}: {str(e)}")
+        import traceback
+        log.severe(f"Traceback: {traceback.format_exc()}")
+
+    sync_duration = time.time() - start_time
+    log.info(
+        f"{entity_name.capitalize()} sync completed. Entities received: {entities_received}, Records processed: {records_processed}. Time taken: {sync_duration:.2f} seconds")
+
+
+def sync_companies(access_token):
+    """
+    Sync company data from Accelo API.
+
+    This function defines the fields to sync for companies and provides
+    a custom record processing function to handle data type conversions.
+    """
+    company_fields = [
+        "id", "name", "website", "phone", "date_created",
+        "date_modified", "date_last_interacted", "comments", "standing", "status",
+        "postal_address", "default_affiliation"
+    ]
+
+    def process_company_record(record):
+        # Convert integer fields
+        int_fields = ["id", "status", "postal_address", "default_affiliation"]
+        for field in int_fields:
+            value = record.get(field)
+            if value is not None:
+                try:
+                    record[field] = int(value)
+                except (ValueError, TypeError):
+                    log.warning(f"Could not convert field {field} with value '{value}' to int")
+                    record[field] = None
+
+        # Convert date fields
```

Comment:
This logic of converting date fields could be a common method. It is same in each method.
`sync_companies`, `sync_payments`, `sync_prospects`, ... etc

Same for integer fields as well.

Reply:
Done